File: LeetCode/Go/374-binary-search-猜数字大小.go
/*
 * @lc app=leetcode.cn id=374 lang=golang
 *
 * [374] 猜数字大小
 */
// package leetcode

// @lc code=start
// /**
//  * Forward declaration of guess API.
//  * @param  num   your guess
//  * @return 	     -1 if num is lower than the guess number
//  *			      1 if num is higher than the guess number
//  *               otherwise return 0
//  * func guess(num int) int;
//  */

func guessNumber(n int) int {
	var low, high = 1, n
	for low <= high {
		var mid = low + (high-low)>>1
		var res = guess(mid)
		if res == 0 {
			return mid
		} else if res < 0 {
			high = mid - 1
		} else {
			low = mid + 1
		}
	}
	return -1
}

// @lc code=end

File: LeetCode/Python/374-binary-search-猜数字大小.py
#
# @lc app=leetcode.cn id=374 lang=python3
#
# [374] 猜数字大小
#
def guess(num: int) -> int:
    return 0


# @lc code=start
# The guess API is already defined for you.
# @param num, your guess
# @return -1 if num is higher than the picked number
#          1 if num is lower than the picked number
#          otherwise return 0
# def guess(num: int) -> int:


class Solution:
    def guessNumber(self, n: int) -> int:
        l, r = 1, n
        while l <= r:
            m = (l + r) // 2
            v = guess(m)
            if v == 0:
                return m
            elif v == 1:
                l = m + 1
            else:
                r = m - 1
        return l


# @lc code=end
